rust/src/ldap/detect.rs

```diff
@@ -401,6 +412,93 @@ unsafe extern "C" fn ldap_tx_get_responses_dn(
     return true;
 }
 
+fn aux_ldap_parse_resp_result_code(s: &str) -> Option<DetectLdapRespResultData> {
+    let parts: Vec<&str> = s.split(',').collect();
+    if parts.len() > 2 {
+        return None;
+    }
+
+    let index = parse_ldap_index(&parts)?;
+    let du32 = detect_parse_uint_enum::<u32, LdapResultCode>(parts[0])?;
+
+    Some(DetectLdapRespResultData { du32, index })
+}
+
+unsafe extern "C" fn ldap_parse_responses_result_code(
+    ustr: *const std::os::raw::c_char,
+) -> *mut DetectUintData<u32> {
+    let ft_name: &CStr = CStr::from_ptr(ustr); //unsafe
+    if let Ok(s) = ft_name.to_str() {
+        if let Some(ctx) = aux_ldap_parse_resp_result_code(s) {
+            let boxed = Box::new(ctx);
+            return Box::into_raw(boxed) as *mut _;
+        }
+    }
+    return std::ptr::null_mut();
+}
+
+unsafe extern "C" fn ldap_detect_responses_result_code_setup(
+    de: *mut c_void, s: *mut c_void, raw: *const libc::c_char,
+) -> c_int {
+    if DetectSignatureSetAppProto(s, ALPROTO_LDAP) != 0 {
+        return -1;
+    }
+    let ctx = ldap_parse_responses_result_code(raw) as *mut c_void;
+    if ctx.is_null() {
+        return -1;
+    }
+    if SigMatchAppendSMToList(
+        de,
+        s,
+        G_LDAP_RESPONSES_RESULT_CODE_KW_ID,
+        ctx,
+        G_LDAP_RESPONSES_RESULT_CODE_BUFFER_ID,
+    )
+    .is_null()
+    {
+        ldap_detect_responses_result_code_free(std::ptr::null_mut(), ctx);
+        return -1;
+    }
+    return 0;
+}
+
+fn get_ldap_result_code(response: &LdapMessage) -> Option<u32> {
+    let result_code = match &response.protocol_op {
+        ProtocolOp::BindResponse(resp) => Some(resp.result.result_code.0),
+        ProtocolOp::SearchResultDone(resp) => Some(resp.result_code.0),
+        ProtocolOp::ModifyResponse(resp) => Some(resp.result.result_code.0),
+        ProtocolOp::AddResponse(resp) => Some(resp.result_code.0),
+        ProtocolOp::DelResponse(resp) => Some(resp.result_code.0),
+        ProtocolOp::ModDnResponse(resp) => Some(resp.result_code.0),
+        ProtocolOp::CompareResponse(resp) => Some(resp.result_code.0),
+        ProtocolOp::ExtendedResponse(resp) => Some(resp.result.result_code.0),
+        _ => None,
+    };
+    return result_code;
+}
+
+unsafe extern "C" fn ldap_detect_responses_result_code_match(
+    _de: *mut c_void, _f: *mut c_void, _flags: u8, _state: *mut c_void, tx: *mut c_void,
+    _sig: *const c_void, ctx: *const c_void,
+) -> c_int {
+    let tx = cast_pointer!(tx, LdapTransaction);
+    let ctx = cast_pointer!(ctx, DetectLdapRespResultData);
+
+    return match_at_index::<u32>(
+        tx,
+        &ctx.du32,
+        get_ldap_result_code,
+        |code, ctx_value| unsafe { rs_detect_u32_match(code, ctx_value) },
```

Comment:
Can we use `detect_match_uint` (maybe `detect_match_uint::<u32>`) to avoid the unsafe ?